### ISSUE TITLE
all_there? assumes all subelements are always present

### ISSUE THREAD

**Issue by tmertens** [natritmeyer/site_prism#48]
The all_there? method expects all elements in a container to always be present.  However, there are often page elements and entire sections which are displayed on the page depending on some condition or action of the user.

Examples:
- Error Message(s)
- Page sections which appear/disappear based on user interaction.
- Modal (Lightbox) Windows
- All descendant sections/elements of aforementioned 'sections'.

It is necessary to have a way of flagging certain elements as optional so that they and any of their descendants do not cause the "all_there?" check against an ancestor element to fail.

Given the page:

``` ruby
class LoginPage < SitePrism::Page
  class LoginForm < SitePrism::Section
    element :password_field, '#password'
    element :username_field, '#username'
    element :submit_button, 'button[type="submit"]'
  end

  class NoticeWindow < SitePrism::Section
    element :notice, ".notice"
    element :alert, ".alert"
  end

  set_url "/login"
  set_url_matcher /login/

  # Page content
  section :login_form, LoginForm, "#loginform"
  element :logo, 'img.logo'
  section :notice_window, NoticeWindow, ".notice-window"
end
```

It is obvious that login_form and logo should be present on page load, and that password_field, username_field, and submit_button should also be present; however, notice_window and/or its subelements should only be present/displayed when a warning or notice must be displayed on the page such as after an invalid login.

So, presuming the required fields described above are present on page load, the following should succeed, but does not:

``` ruby
page = LoginPage.new
page.load
page.should be_all_there #=> all_there? returns false and fails validation
```

Do you have any thoughts on how to proceed with a change to implement this?  I've implemented such a mechanism in a testing framework I worked on previously; however, each element was its own object with its own 'required' flag, whereas in SitePrism, individual elements are simply instantiated as a method call to Capybara and held in the containing section/page as a string in the mapped_items array.

IMO each element should be its own object and respond to `element#required?`, which defaults to true but can be set to false for stateful elements.

**Comment by luke-hill** [comment]
Hi @tmertens You'll notice I've been bumping a few tickets recently and getting peoples input.

Can this and #49 be looked at and closed. I believe there is a PR tidying up `all_there?` slightly which will be merged soon.

**Comment by luke-hill** [comment]
Closing as fixed - Released in site_prism `v2.11`
